cirq/circuits/circuit.py

```diff
@@ -1258,6 +1258,60 @@ def _json_dict_(self):
     def _from_json_dict_(cls, moments, device, **kwargs):
         return cls(moments, strategy=InsertStrategy.EARLIEST, device=device)
 
+    def raggedy_add(*circuits: 'cirq.Circuit') -> 'cirq.Circuit':
+        """Concatenates circuits while overlapping them if possible.
+
+        Starts with the first circuit (index 0), then iterates over the other
+        circuits while folding them in. To fold two circuits together, they
+        are placed one after the other and then the second one is moved backward
+        until either just before its operations would collide with operations
+        in the first one or else their ends are equal.
```

Comment:
Is it important that their ends stay equal if there's no collision? Otherwise you could keep moving the second circuit backward until there's actually a collision, which, if you're adding multiple circuits together, could end up more compact overall. It would also preserve associativity this way.

Reply:
Hmm, I definitely conceived of this method as trying to line up the starts or the ends as opposed to shoving it as far in as possible. Maybe it should be a parameter.

Reply:
I'd pick whichever fits your use case and yagni the parameter. Mainly I came to this while thinking about the associativity. But ultimately associativity is not important.

Reply:
Continuing to move it backward doesn't restore associativity. I updated the example to reflect this.

The method now has `stop_at_first_alignment = False` as a parameter. By default it stops at the last alignment. There's no option for continuing past that point when the circuits act on disjoint qubits, because I havea hard time picturing why you would ever want the second one to end up in front of the first one.

Reply:
Ah, I didn't consider those examples. Yeah I think that parameter makes sense then.